File: fm_api/schemas/shopping_list.py
from pydantic import BaseModel
from typing import Optional, List


class ShoppingListRequest(BaseModel):
    recipe_ids: List[int] = []
    ingredient_ids: List[int] = []


class ShoppingListItem(BaseModel):
    ingredient_id: int
    name: str
    quantity: Optional[float] = None
    unit: Optional[str] = None
    source_recipes: List[str] = []


class UnlinkedItem(BaseModel):
    name: str
    quantity: Optional[float] = None
    unit: Optional[str] = None
    recipe_name: str


class ShoppingListResponse(BaseModel):
    needed: List[ShoppingListItem] = []
    unlinked: List[UnlinkedItem] = []

File: fm_frontend/src/api/shopping_list.ts
import { apiFetch } from './client'

export interface ShoppingListItem {
  ingredient_id: number
  name: string
  quantity: number | null
  unit: string | null
  source_recipes: string[]
}

export interface UnlinkedItem {
  name: string
  quantity: number | null
  unit: string | null
  recipe_name: string
}

export interface ShoppingListResponse {
  needed: ShoppingListItem[]
  unlinked: UnlinkedItem[]
}

export async function generateShoppingList(
  recipe_ids: number[],
  ingredient_ids: number[]
): Promise<ShoppingListResponse> {
  return apiFetch('/shopping-list/generate', {
    method: 'POST',
    body: JSON.stringify({ recipe_ids, ingredient_ids }),
  })
}
